Execute the Windows installer as normal user

mardukbp opened the pull request:
Currently the Windows installer requires admin privileges, which are typically not available in an enterprise environment. This PR produces a Windows installer that runs without admin privileges.

Three of the four reasons why the installer requires admin rights have been addressed: 

- HKCR registry keys were replaced by HKCU registry keys
- The name of the Erlang node of the Livebook app includes the host IP (since name resolution apparently requires admin rights)
- The vc_redist.x64.exe installer was removed and the Livebook.exe app no longer targets .NET Framewok 4. Now it targets .NET 8 and is completely self-contained.

OTP (or rather erl.exe) requires vcredist140.dll. This DLL can be added after extracting the OTP distribution.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/livebook-dev/livebook?pullRequest=2628) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/livebook-dev/livebook?pullRequest=2628) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/livebook-dev/livebook?pullRequest=2628) it.</sub>

wojtekmach commented:
> OTP (or rather erl.exe) requires vcruntime140.dll. This DLL can be added after extracting the OTP distribution.

Can you elaborate on this? If we still need .dll but we are not shipping its installer and thus not installing it ourselves, how does this work in practice?

mardukbp commented:
I just learned that the [windows-latest](https://github.com/actions/runner-images/blob/main/images/windows/Windows2022-Readme.md) image of GH actions includes VS 2022 with the component Microsoft.VisualStudio.Component.VC.Redist.14.Latest. And I found [a script](https://github.com/Z3Prover/z3/blob/904a50fe498c325f5c347628cd66a1ab610030e0/scripts/mk_win_dist.py#L264) that copies the DLL from it.

Python, VS Code and many Rust CLI apps ship this DLL.

wojtekmach commented:
Sorry I still don’t understand how it works. If we are not shipping the .dll how does Livebook work without it? Are you saying users would have gotten it after installing, say, VSCode? But then it wouldn’t work without it, right?

wojtekmach commented:
Are you saying we should ship .dll ourselves (as opposed to shipping the vc_redist.exe which installs it into location which requires admin)? If so, yes, that’d be great and help on that front would be very appreciated.

mardukbp commented:
Yes, that's what I mean. I will ask the Erlang people why don't they do it themselves. This issue leaves students and people in companies out of the BEAM world. If for some reason they don't want to do it, I will implement it for Livebook :)

wojtekmach commented:
Right, thanks.

So about the other changes. Switching to HKCU sounds good to me.

About dropping Windows.Forms and .NET 4.6 with self-contained=false in favor of .NET 8.0 with self-contained=true and using third party plugins, what is the installer size difference? Could you list upsides/downsides of what you are proposing? In isolation from other issues (vc redist and HKCU) because I believe, and correct me if I’m wrong, it is completely orthogonal?

wojtekmach commented:
The PR looks good to me assuming OTP will ship with required .dll files. If you could start the conversation and/or PR and if accepted it would be released in OTP 27.1, I think we can wait. Otherwise, eg it is deamed to big of a change and so scheduled for OTP 28, we’d appreciate doing the change in Livebook.

wojtekmach commented:
If you start issue/PR in OTP please cc me and/or mention it here so we can keep track!

mardukbp commented:
I opened an [issue ](https://github.com/erlang/otp/issues/8531)in the OTP repo. Subscribe and we'll see what they say :)

mardukbp commented:
Hi @wojtekmach, I just updated the installer so that uninstalling works. I also reverted the removal of vc_redist.exe. 

The Livebook installer would work for anybody who already has vcruntime140.dll. Hopefully that is a large percentage of users. And hopefully OTP will one day ship the DLL. Then Livebook will not need vc_redist.exe at all.

wojtekmach commented:
Could you rebase with latest main? I’d expect your changes around node naming and similar are no longer necessary.

mardukbp commented:
Done

mardukbp commented:
I tested it with a user with no admin rights and Livebook does not work. The notebook cannot connect to Elixir. Something is missing in the new env.bat. I will have to investigate, unless you have an idea.

mardukbp commented:
@wojtekmach I got it working :) Since I am behind a VPN, following the Installation FAQ I created  ".livebookdesktop.bat" in my User Profile folder with the contents:

```
set LIVEBOOK_DISTRIBUTION=name
set LIVEBOOK_NODE=livebook@127.0.0.1
```

My suggestion would be that the installer creates this file. Unless it can negatively affect some users.

josevalim commented:
I believe those are the default in latest Livebook, so we should be good to go!

mardukbp commented:
I built the installer and tested it on a Windows machine with no admin rights. Livebook is now enterprise-ready!

wojtekmach reviewed (approved):
I tested this and it works great.

There's a minor caveat that the vcredist installer shows up every single time:

<img width="526" alt="image" src="https://github.com/user-attachments/assets/65062482-15db-4a02-85fe-4eb5bf4f8904">

My guess is we didn't see it previously because the installer was running as an admin and maybe then they don't bother asking.

I think I can easily fix this by only conditionally running the vcredist installer when needed (by checking appropriate winreg keys)

Thanks again!

wojtekmach commented:
When we release it, we need to tell people that if they previously installed Livebook, they need to go to "Settings / Apps / Installed Apps" and manually uninstall it first. I'm a little bit worried people might run into weird issues where previously the installer was running as an admin and now as a user. I tested it though, installing nightly, uninstalling it, installing this branch, and everything works well.

@josevalim @jonatanklosko should we merge this or wait until after v0.15 is out? I'm keen on users ideally trying this out with nightlies and reporting any issues.

josevalim commented:
@wojtekmach could we perhaps use another directory, to avoid the admin conflict issue? In any case, feel free to merge this any time.

wojtekmach commented:
You know what, I think this is probably fine and if people run into issues we'd tell them to uninstall and try again anyway.

hugobarauna commented:
@wojtekmach 

> @josevalim @jonatanklosko @hugobarauna should we merge this or wait until after v0.15 is out? I'm keen on users ideally trying this out with nightlies and reporting any issues sooner than later.

No comments on this one. Defer to José and Jonatan.

That said, I see it's already merged. So, great! :)

mardukbp commented:
@wojtekmach Thanks for all the time invested in reviewing and testing this PR. As I have mentioned before, the definitive solution for vcredist is including that one DLL in the distribution. Since it is actually needed by Erlang, they should do it. I already opened an issue in their repo and argued that requiring admin rights for installing Erlang leaves people at enterprises and universities out of the BEAM ecosystem.

 I won't be at Code Beam Europe, but maybe you or @josevalim can talk to them about this :)
